pkg/kv/kvserver/flow_control_integration_test.go

```diff
@@ -4917,6 +4917,84 @@ ORDER BY streams DESC;
 	h.query(n1, flowPerStoreTokenQueryStr, flowPerStoreTokenQueryHeaderStrs...)
 }
 
+func TestFlowControlRepeatedlySwitchMode(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	settings := cluster.MakeTestingClusterSettings()
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+		ServerArgs: base.TestServerArgs{
+			Settings: settings,
+			RaftConfig: base.RaftConfig{
+				// Suppress timeout-based elections. This test doesn't want to
+				// deal with leadership changing hands.
+				RaftElectionTimeoutTicks: 1000000,
+			},
+			Knobs: base.TestingKnobs{
+				Store: &kvserver.StoreTestingKnobs{
+					FlowControlTestingKnobs: &kvflowcontrol.TestingKnobs{
+						UseOnlyForScratchRanges: true,
+						OverrideTokenDeduction: func(_ kvflowcontrol.Tokens) kvflowcontrol.Tokens {
+							// This test makes use of (small) increment requests, but
+							// wants to see large token deductions/returns.
+							return kvflowcontrol.Tokens(1 << 20 /* 1MiB */)
+						},
```

Comment:
What does this do? Is this to a) workaround the fact that we don't know exactly the size of proposals, and/or b) we want the streams to be blocked earlier without sending thousands of proposals?

Does it apply both to eval and send tokens?

Reply:
a) yes and we cannot construct an exact sized proposal directly for all sizes
b) also yes, we could issue a 16 MiB write but we've observed that to be extremely flaky under stressrace due to overload affecting the test
<!-- Reviewable comment -OC4onuhAjjFjD2sjsBV:b253f8e -->
<!-- Sent from Reviewable.io -->